snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,300 @@
+from snuba.request.request_settings import RequestSettings
+
+tags_key_pattern = FunctionCall(
+    None, String("arrayJoin"), (Column(None, None, String("tags.key")),)
+)
+
+
+def _get_tag_keys_in_condition(condition: Expression) -> Optional[Set[str]]:
+    """
+    Finds the top level conditions that include filter based on tags_key.
+    This is meant to be used to find the tag keys the query is filtering on.
+    We can only apply the arrayFilter optimization to tag keys conditions
+    that are not in OR with other columns. To simplify the problem, we only
+    consider those conditions that are included in the first level of the query:
+    [['tagskey' '=' 'a'],['col' '=' 'b'],['col2' '=' 'c']]  works
+    [[['tagskey' '=' 'a'], ['col2' '=' 'b']], ['tagskey' '=' 'c']] does not
+
+    If we encounter an OR condition we return None, which means we cannot
+    safely apply the optimization. Empty set means we did not find any
+    suitable tag key for optimization in this condition but that does
+    not disqualify the whole query in the way the OR condition does.
+    """
+    tag_keys_found = set()
+
+    conditions = get_first_level_conditions(condition)
+    for c in conditions:
+        if is_binary_condition(c, BooleanFunctions.OR):
+            return None
+
+        match = FunctionCall(
+            None,
+            String(ConditionFunctions.EQ),
+            (tags_key_pattern, Literal(None, Param("key", Any(str)))),
+        ).match(c)
+        if match is not None:
+            tag_keys_found.add(match.string("key"))
+
+        match = is_in_condition_pattern(tags_key_pattern).match(c)
+        if match is not None:
+            function = cast(FunctionCallExpr, match.expression("tuple"))
+            tag_keys_found |= {
+                l.value
+                for l in function.parameters
+                if isinstance(l, LiteralExpr) and isinstance(l.value, str)
+            }
+
+    return tag_keys_found
+
+
+def get_filtered_tag_keys(query: Query) -> Set[str]:
+    """
+    Identifies the tag names we can apply the arrayFilter optimization on.
+    Which means: if the tags_key column is in the select clause, there
+    are one or more top level AND condition on the tags_key column and
+    there is no OR condition in the query.
+    """
+    tags_key_found = any(
+        tags_key_pattern.match(f) is not None
+        for expression in query.get_selected_columns_from_ast() or []
+        for f in expression
+    )
+
+    if not tags_key_found:
+        return set()
+
+    ast_condition = query.get_condition_from_ast()
+    cond_tags_key = (
+        _get_tag_keys_in_condition(ast_condition)
+        if ast_condition is not None
+        else set()
+    )
+    if cond_tags_key is None:
+        # This means we found an OR. Cowardly we give up even though there could
+        # be cases where this condition is still optimizable.
+        return set()
```

Comment:
I don't quite get why we return the empty set here and totally skip the attempt to perform the optimization on HAVING. And vice versa. Wouldn't that still be valid / unrelated to whether or not we found an OR here?

Reply:
We cannot guarantee the query would be correct.
This is the same logic we execute today:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L243-L245

The issue is that if we are using tags_key and tags_value together in the same query, the arrayJoin would happen only once and then we would use the alias in the other cases. If the `WHERE` condition can be optimized but the `HAVING` cannot, and we optimized the `WHERE`, then we would end up using the same condition in the having (the `all_tags` alias), which would be wrong.

So this takes the conservative approach and just bail in case it is not 100% sure the condition is optimizable at the cost of having false negatives (conditions that could be optimized but that are not).